### PR TITLE
Support ? Kleene macro operator in 2015

### LINKED ISSUE #56668: Extending `?` macro repetition to Rust 2015
We recently stabilized the `?` macro repeater "operator" proposed by @mark-i-m for Rust 2018 and it will be in Rust 1.32.

The original https://github.com/rust-lang/rfcs/pull/2298 did not say anything about using edition dependent behavior; this was instead decided upon with an FCP in https://github.com/rust-lang/rust/issues/51934. This was done after a crater run in https://github.com/rust-lang/rust/pull/49719#issuecomment-381364742 that was green and showed 0 regressions.

After accepting the edition dependent change, @petrochenkov noted in https://github.com/rust-lang/rust/issues/51934#issuecomment-402537350 that:
> As you can see in #51587 the edition-specific logic is much worse than previously implemented fallback.

However, in https://github.com/rust-lang/rust/issues/53668, there was a similar situation of a syntactic change for non-soundness reasons that was theoretically breaking but in practice not. In that case, @rust-lang/lang decided to accept the change for Rust 2015 and 2018 in an effort to minimize technical debt.

In #53668 @petrochenkov noted that (https://github.com/rust-lang/rust/issues/53668#issuecomment-423144713):

>  Changes with negligible effects on contrived code are done very regularly, it's just nobody notices them.

In https://github.com/rust-lang/rust/issues/53668#issuecomment-423319580, @nikomatsakis then expanded upon the practical way in which we interpret breaking changes.

If I recall correctly, in a language team meeting not too long ago, it was suggested that we might want to reverse our decision and extend `?` as a macro repetition "operator" in Rust 2015 as well. I think personally that it would be advisable; in particular, if we do so, we can avoid the technical debt in the compiler, the debt in the spec, but first and foremost, we can avoid the distinction that users must learn. Ultimately I think this is of greater practical benefit than the theoretical breakage that it would be according to crater.

Therefore I'd like to use this issue to propose an extension of `?` as a macro repetition "operator" to Rust 2015.

@mark-i-m  noted that they would rather not have to change the implementation again so I will not ask them to; but I'm sure someone else might want to tackle it (e.g. @petrochenkov since they were one of the first proponents of keeping it in Rust 2015 or @alexreg since they concurred with @petrochenkov).

Thoughts?

### PR DESCRIPTION
**Pull request by Centril**
Closes https://github.com/rust-lang/rust/issues/56668.

See that issue for rationale and discussion.

Crater will be needed (done in https://github.com/rust-lang/rust/pull/60932#issuecomment-494188577, zero regressions) and then, if all goes well, FCP (in https://github.com/rust-lang/rust/pull/60932#issuecomment-494189802).

### CODE REVIEW

**Comment by Centril** [comment]
@craterbot run mode=check-only

**Comment by Centril** [comment]
r? @petrochenkov for actual code review.

@rfcbot merge

Per note in https://github.com/rust-lang/rust/issues/56668#issuecomment-447114860 we said that we would do this if crater said it didn't break anyone and indeed it does not. Therefore, I propose that we extend support for `?` as the Kleene at-most-one macro operator to 2015. The rationale is provided in https://github.com/rust-lang/rust/issues/56668#issue-389196697. The gist of it is that we want to reduce technical debt in the compiler, the spec, and also reduce edition differences that must be learned. The original stabilization report summarizing the behavior on 2018 can be found in https://github.com/rust-lang/rust/issues/48075#issuecomment-433243048.

**Comment by Centril** [comment]
N.B. Aaron is on leave so I've checked his box.

File: src/libsyntax/ext/tt/quoted.rs [hunk 1]
```diff
@@ -411,164 +401,11 @@ where
...
-    _features: &Features,
-    _attrs: &[ast::Attribute],
-) -> (Option<token::Token>, KleeneOp)
-where
-    I: Iterator<Item = tokenstream::TokenTree>,
-{
+) -> (Option<token::Token>, KleeneOp) {
     // We basically look at two token trees here, denoted as #1 and #2 below
     let span = match parse_kleene_op(input, span) {
         // #1 is a `?` (needs feature gate)
```
Comment (mark-i-m) [inline src/libsyntax/ext/tt/quoted.rs]: Outdated comment.

File: src/test/run-pass/macros/macro-at-most-once-rep-2015.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// run-pass
+#![allow(unused_mut)]
+// The logic for parsing Kleene operators in macros has a special case to disambiguate `?`.
+// Specifically, `$(pat)?` is the ZeroOrOne operator whereas `$(pat)?+` or `$(pat)?*` are the
+// ZeroOrMore and OneOrMore operators using `?` as a separator. These tests are intended to
+// exercise that logic in the macro parser.
```
Comment (mark-i-m) [inline src/test/run-pass/macros/macro-at-most-once-rep-2015.rs]: This is not true any more. `?` is not allowed as a separator. `$(pat)?+` matches `pat +` or `+`, _not_ `pat`, `pat ? pat`, etc

File: src/test/run-pass/macros/macro-at-most-once-rep-2015.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+//
+// Moreover, we also throw in some tests for using a separator with `?`, which is meaningless but
+// included for consistency with `+` and `*`.
```
Comment (mark-i-m) [inline src/test/run-pass/macros/macro-at-most-once-rep-2015.rs]: This is also not true. Using a separator with `?` will result in a parsing error.

Comment (mark-i-m) [reply src/test/run-pass/macros/macro-at-most-once-rep-2015.rs]: It looks like I just forgot to update the comments the last time I updated this file. The test itself looks correct to me...

**Review by mark-i-m — APPROVED** [review]
Code LGTM, but there are a few outdated comments.

**Comment by mark-i-m** [comment]
r=me after FCP and with comments updated :+1:

**Comment by Centril** [comment]
@mark-i-m Adjusted per review comments & did some cleanup.

**Review by mark-i-m — APPROVED** [review]
r=me with FCP

**Comment by mark-i-m** [comment]
Just learn about this... eager to use it :laughing: 

@rfcbot modify labels: -S-waiting-on-review +S-waiting-on-team

**Comment by Centril** [comment]
...rebased

@bors r=mark-i-m

**Pull request merged**